File: 19_RemoveNthNodeFromEndOfList/19_v1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        vector<ListNode*> v;
        for (ListNode* p = head; p != nullptr; p = p->next)
            v.push_back(p);
        
        int idx = v.size() - n;
        ListNode *removeNode = v[idx];
        
        if (idx == 0){
            head = removeNode->next;
        } else {
            ListNode *prevNode = v[idx - 1];
            prevNode->next = removeNode->next;
        }        

        return head;
    }
};

File: 19_RemoveNthNodeFromEndOfList/19_v1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ArrayList<ListNode> l = new ArrayList<ListNode>();
        for (ListNode p = head; p != null; p = p.next)
            l.add(p);
            
        int idx = l.size() - n;
        ListNode removeNode = l.get(idx);
        
        if (idx == 0) {
            head = removeNode.next;
        } else {
            ListNode prevNode = l.get(idx - 1);
            prevNode.next = removeNode.next;
        }
        
        return head;
    }
}
